trieste/acquisition/rule.py

```diff
@@ -211,31 +212,208 @@ def acquire(
+
+        :param points: Points to remove.
+        :return: New instance of `AsynchronousRuleState` with updated pending points.
+        """
+        if not self.has_pending_points:
+            # nothing to do if there are no pending points
+            return self
+
+        tf.debugging.assert_shapes(
+            [(self.pending_points, [..., "D"]), (points, [..., "D"])],
+            message=f"""Last dimension mismatch between pending points
+                        {tf.shape(self.pending_points)} and given points {tf.shape(points)}""",
+        )
+
+        @tf.function
+        def is_not_in_points(x: TensorType) -> TensorType:
+            return tf.reduce_all(tf.reduce_any(tf.not_equal(points, x), axis=1))
+
+        mask = tf.map_fn(is_not_in_points, self.pending_points, fn_output_signature=tf.bool)
+        return AsynchronousRuleState(tf.boolean_mask(self.pending_points, mask))
+
+    def add_points(self, new_points: TensorType) -> AsynchronousRuleState:
+        """Adds `new_points` to the already known pending points.
+
+        :param new_points: Points to add.
+        :return: New instance of `AsynchronousRuleState` with updated pending points.
+        """
+        if not self.has_pending_points:
+            return AsynchronousRuleState(new_points)
+
+        tf.debugging.assert_shapes(
+            [(self.pending_points, [..., "D"]), (new_points, [..., "D"])],
+            message=f"""Last dimension mismatch between pending points
+                        {tf.shape(self.pending_points)} and given points {tf.shape(new_points)}""",
+        )
+
+        new_pending_points = tf.concat([self.pending_points, new_points], axis=0)
+        return AsynchronousRuleState(new_pending_points)
+
+
+class AsynchronousOptimization(
+    AcquisitionRule[State[Optional["AsynchronousRuleState"], TensorType], SP_contra]
 ):
-    """AsynchronousGreedy rule, as name suggests,
-    is designed for asynchronous BO scenarios.
+    """AsynchronousOptimization rule is designed for asynchronous BO scenarios.
     By asynchronous BO we understand a use case when multiple objective function
     can be launched in parallel and are expected to arrive at different times.
     Instead of waiting for the rest of observations to return, we want to immediately
     use acquisition function to launch a new observation and avoid wasting computational resources.
 
     To make the best decision about next point to observe, acquisition function
     needs to be aware of currently running observations.
-    We call such points "pending". AsyncEGO provides a ``AsynchronousGreedy.State``
-    object that keeps track of pending points.
+    We call such points "pending", and consider them a part of acquisition state.
+    We use :class:`AsynchronousRuleState` to store these points.
+
+    AsynchronousOptimization works with non-greedy batch acquisition functions.
+    If there are P pending points, the acquisition function is used with batch size P+1.
```

Comment:
This was also discussed in previous PR, where we actually removed batch size parameter. There is no real use for it in async scenarios, you can just ask for the next point as many time as you'd like and all previous pending points will be taken into account. Also, normally you ask for a point because your worker just finished working on the previous observation, so there is no need for more than one anyway.